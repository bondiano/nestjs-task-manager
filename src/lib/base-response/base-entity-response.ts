import { ApiProperty } from '@nestjs/swagger';

import { IBaseEntityData } from '../types/response';

import { BaseResponse } from './base-response';

export class BaseEntityResponse<Data>
  extends BaseResponse<Data>
  implements IBaseEntityData
{
  @ApiProperty({ format: 'int32', required: true })
  id: number;

  @ApiProperty({ required: true })
  createdAt: number;

  @ApiProperty({ required: true })
  updatedAt: number;

  @ApiProperty({ required: false })
  deletedAt: number | null;

  constructor(data: Data) {
    super(data);
  }
}
