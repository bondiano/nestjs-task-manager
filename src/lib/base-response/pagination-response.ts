import { Constructor } from 'type-fest';

import {
  ApiProperty,
  ApiResponseSchemaHost,
  getSchemaPath,
} from '@nestjs/swagger';

import { IPaginationOptions, IPaginationData } from '../types/response';

export class PageMetaDto implements IPaginationOptions {
  @ApiProperty()
  public readonly count: number;

  @ApiProperty()
  public readonly total: number;

  @ApiProperty()
  public readonly perPage: number;

  @ApiProperty()
  public readonly currentPage: number;

  @ApiProperty()
  public readonly totalPages: number;

  @ApiProperty()
  public readonly searchParams: Record<string, unknown>;

  constructor(options: IPaginationOptions) {
    this.count = options.count ?? 0;
    this.total = options.total ?? 0;
    this.perPage = options.perPage ?? 0;
    this.currentPage = options.currentPage ?? 1;
    this.totalPages = options.totalPages ?? 1;
    this.searchParams = options.searchParams ?? {};
  }
}

export class PaginationResponse<Data> implements IPaginationData<Data> {
  /**
   * Takes ApiModel or [ApiModel] (class with swagger-decorated properties).
   * Returns compiled SchemaObject for array.
   * Don't forget to add @ApiExtraModels(ClassName) to controller.
   */
  static buildOpenApiSchema<T extends string | Constructor<unknown>>(
    entity: T
  ): ApiResponseSchemaHost['schema'] {
    return {
      type: 'object',
      properties: {
        success: {
          type: 'boolean',
        },
        payload: {
          type: 'object',
          properties: {
            data: {
              type: 'array',
              items: {
                $ref: getSchemaPath(entity),
              },
            },
            meta: {
              $ref: getSchemaPath(PageMetaDto),
            },
          },
        },
      },
    };
  }

  static buildPagerOptions<T>(
    items: Array<unknown>,
    total: number,
    pager?: {
      page?: number;
      limit?: number;
      searchParams?: Partial<Record<keyof T, unknown>>;
    }
  ): Required<IPaginationOptions> {
    const _pager = {
      page: pager?.page ?? 1,
      limit: pager?.limit ?? items.length,
    };

    const totalPages = Math.ceil(total / _pager.limit);

    return {
      total,
      totalPages: Number.isNaN(totalPages) ? 1 : totalPages,
      count: items.length,
      perPage: _pager.limit,
      currentPage: _pager.page,
      searchParams: pager?.searchParams ?? {},
    };
  }

  @ApiProperty({
    type: [],
  })
  public readonly items: Array<Data>;

  @ApiProperty()
  public readonly meta: PageMetaDto;

  constructor(items: Array<Data>, options: IPaginationOptions) {
    this.items = items;
    this.meta = new PageMetaDto(options);
  }
}
