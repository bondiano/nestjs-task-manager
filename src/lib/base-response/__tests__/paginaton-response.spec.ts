import { ApiProperty } from '@nestjs/swagger';

import { BaseEntityResponse } from '../base-entity-response';
import { PaginationResponse } from '../pagination-response';

describe('PaginationResponse', () => {
  class PostResponse extends BaseEntityResponse<{ title: string }> {
    @ApiProperty()
    title: string;
  }

  it('should build correct schema', () => {
    const schema = PaginationResponse.buildOpenApiSchema(PostResponse);

    expect(schema).toMatchSnapshot();
  });

  it('should define correct properties', () => {
    const paginationResponse = new PaginationResponse([], {});

    expect(paginationResponse).toHaveProperty('items');
    expect(paginationResponse).toHaveProperty('meta');

    expect(paginationResponse.meta.count).toBe(0);
    expect(paginationResponse.meta.total).toBe(0);
    expect(paginationResponse.meta.perPage).toBe(0);
    expect(paginationResponse.meta.currentPage).toBe(1);
    expect(paginationResponse.meta.totalPages).toBe(1);
  });

  it('should work with BaseResponse', () => {
    const paginationResponse = new PaginationResponse<{ title: string }>(
      PostResponse.fromArray([{ title: 'title' }]),
      {
        count: 1,
        total: 1,
        perPage: 1,
        currentPage: 1,
      }
    );

    expect(paginationResponse).toHaveProperty('items');
    expect(paginationResponse).toHaveProperty('meta');
    expect(paginationResponse.meta.totalPages).toBe(1);
  });

  it('should build correct options', () => {
    const posts = PostResponse.fromArray([
      { title: 'test2' },
      { title: 'test1' },
      { title: 'test' },
    ]);

    const paginationResponse = new PaginationResponse(
      posts,
      PaginationResponse.buildPagerOptions(posts, 101, {
        page: 1,
        limit: 3,
        searchParams: {},
      })
    );

    expect(paginationResponse).toHaveProperty('items');
    expect(paginationResponse).toHaveProperty('meta');
    expect(paginationResponse.meta.totalPages).toBe(34);
    expect(paginationResponse.meta.count).toBe(3);
    expect(paginationResponse.meta.total).toBe(101);
  });
});
