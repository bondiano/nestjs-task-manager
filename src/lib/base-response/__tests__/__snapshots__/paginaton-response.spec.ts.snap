// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationResponse should build correct schema 1`] = `
{
  "properties": {
    "payload": {
      "properties": {
        "data": {
          "items": {
            "$ref": "#/components/schemas/PostResponse",
          },
          "type": "array",
        },
        "meta": {
          "$ref": "#/components/schemas/PageMetaDto",
        },
      },
      "type": "object",
    },
    "success": {
      "type": "boolean",
    },
  },
  "type": "object",
}
`;
